wallet, mempool: propagete `checkChainLimits` error message to wallet

ismaelsadeeq opened the pull request:
* Requested in [#28391 comment](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1382997719)

* The error message is static when a new transaction is created and package limit is reached.
`Transaction has too long of a mempool chain`
While the [`CTxMempool::CheckPackageLimits`](https://github.com/bitcoin/bitcoin/blob/5800c558eb5efb4839ed00d6967e43306d68e1c3/src/txmempool.cpp#L199) provide explicit information about the error message.
* This PR updates [`CTxMempool::CheckPackageLimits`](https://github.com/bitcoin/bitcoin/blob/5800c558eb5efb4839ed00d6967e43306d68e1c3/src/txmempool.cpp#L199) return type to `util::Result<void>`, `CheckPackageLimits` now returns void when package limit is not hit, and returns the error string whenever package limit is hit instead of using out parameter `errString`.
* The PR updates [`checkChainLimits`](https://github.com/bitcoin/bitcoin/blob/5800c558eb5efb4839ed00d6967e43306d68e1c3/src/node/interfaces.cpp#L703) return type to `util::Result<void>`.

* Now the wallet `CreateTransactionInternal` will have access to the package limit error string whenever its hit.
* Also Updated functional test to reflect the error message from `CTxMempool::CheckPackageLimits` output.

Sjors commented on test/functional/wallet_basic.py:
```diff
@@ -612,10 +612,11 @@ def run_test(self):
         sending_addr = self.nodes[1].getnewaddress()
         txid_list = []
-        for _ in range(chainlimit * 2):
+        chain_limit_2x = 2 * chainlimit
```
nit: I don't think adding this variable improves readability

Sjors reviewed (commented):
Concept ACK.

It would be nice if the test can reach all four possible limit error messages, but if that's difficult to achieve it could wait for a followup.

ismaelsadeeq commented:
> It would be nice if the test can reach all four possible limit error messages, but if that's difficult to achieve it could wait for a followup.

Previosuly we always got the `Transaction has too long of a mempool chain` message but with this PR we now get the node unconfirmed parents limit.
[checkChainLimits](https://github.com/bitcoin/bitcoin/blob/5800c558eb5efb4839ed00d6967e43306d68e1c3/src/node/interfaces.cpp#L703) is sending a package with one Internally created transaction and its `vsize` as the package `vsize`, package count will always be one.

BrandonOdiwuor reviewed (commented):
Concept ACK 

using the error message from `checkChainLimits(...)`

BrandonOdiwuor commented on src/node/interfaces.cpp:
```diff
@@ -58,6 +58,7 @@
 #include <any>
 #include <memory>
 #include <optional>
+#include <string>
```
Nit: I believe this addition might be unnecessary, considering that other instances of `std::string` in the file seem to work fine without it

BrandonOdiwuor reviewed (commented):
The PR looks good, left a Nit comment

Sjors commented:
Note that with cluster mempool #28676 the error would change. Which is another reason why I think it's useful to have tests cases for all four scenario's. But it can wait for another PR, and perhaps @sdaftuar could write it :-)

ismaelsadeeq commented:
Thanks @Sjors 
I Would also like to work on it whenever that happens.

sedited reviewed (approved):
ACK bd5417e47cd2452817d7de859839b63210233a5c

furszy commented on src/interfaces/chain.h:
```diff
@@ -260,7 +260,7 @@ class Chain
 
     //! Check if transaction will pass the mempool's chain limits.
-    virtual bool checkChainLimits(const CTransactionRef& tx) = 0;
+    virtual bool checkChainLimits(const CTransactionRef& tx, std::string& err_string) = 0;
```
Not blocking.
Could return `util::Result<void>` and inline all the `err_string` return messages within the `CheckPackageLimits` function.

ismaelsadeeq replied:
Thanks  @furszy updated to your suggestion, its more elegant, and no for out parameter `err_string`

ismaelsadeeq committed:
wallet, mempool: propagete `checkChainLimits` error message to wallet

Update CheckPackageLimits to use util::Result to pass the error message
instead of out parameter.

Also update test to reflect the error message from `CTxMempool`
`CheckPackageLimits` output.

ismaelsadeeq commented:
Thanks for review @TheCharlatan @furszy 

- Addressed @furszy comment
- Forced pushed bd5417e47cd2452817d7de859839b63210233a5c to 28115d81b2 [Compare diff](https://github.com/bitcoin/bitcoin/compare/bd5417e47cd2452817d7de859839b63210233a5c..28115d81b2eacb15dd4835d5536a71b0b5ac835b)

BrandonOdiwuor reviewed (approved):
ACK 28115d81b2eacb15dd4835d5536a71b0b5ac835b

TheCharlatan commented on src/node/interfaces.cpp:
```diff
@@ -698,14 +699,15 @@ class ChainImpl : public Chain
         LOCK(m_node.mempool->cs);
         std::string err_string;
-        return m_node.mempool->CheckPackageLimits({tx}, entry.GetTxSize(), err_string);
+        if (m_node.mempool->CheckPackageLimits({tx}, entry.GetTxSize(), err_string)) return {};
```
Since we're now changing interfaces, I think `CheckPackageLimits` could return a `Result` type too.

ismaelsadeeq replied:
Taken in https://github.com/bitcoin/bitcoin/pull/28863/commits/e572a530d224214121371c9e42a013ef3b26ee0e
Thanks

ismaelsadeeq commented:
Force pushed from https://github.com/bitcoin/bitcoin/commit/28115d81b2eacb15dd4835d5536a71b0b5ac835b to https://github.com/bitcoin/bitcoin/commit/e572a530d224214121371c9e42a013ef3b26ee0e 
[Compare diff](https://github.com/bitcoin/bitcoin/compare/28115d81b2eacb15dd4835d5536a71b0b5ac835b..e572a530d224214121371c9e42a013ef3b26ee0e)
* Updated the PR OP and title
* Addressed @TheCharlatan comment https://github.com/bitcoin/bitcoin/pull/28863#discussion_r1404640801

TheCharlatan commented on src/wallet/spend.cpp:
```diff
@@ -1251,8 +1251,9 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
-            return util::Error{_("Transaction has too long of a mempool chain")};
+        auto result = wallet.chain().checkChainLimits(tx);
+        if (!result) {
+            return util::Error{util::ErrorString(result)};
```
Just a comment: Would be nice to get #25665 merged, so we can get a move constructor here.

ismaelsadeeq replied:
Thanks @TheCharlatan that can come as a follow-up post https://github.com/bitcoin/bitcoin/pull/25665 I think?

Sjors replied:
Indeed, or it could even be done as part of that PR. It depends on which one gets merged first.

Sjors commented on src/txmempool.h:
```diff
@@ -605,11 +605,9 @@ class CTxMemPool
      *                                          to mempool. The transactions need not be direct
      *                                          ancestors/descendants of each other.
      * @param[in]       total_vsize             Sum of virtual sizes for all transactions in package.
-     * @param[out]      errString               Populated with error reason if a limit is hit.
```
```cpp
* @returns {} or the error reason if a limit is hit.
```

instagibbs replied:
this would be good

ismaelsadeeq replied:
Thanks Would create a tiny follow-up to fix this, so as not to invalidate ACK's.

ismaelsadeeq replied:
This is fixed in https://github.com/bitcoin/bitcoin/pull/29115

Sjors reviewed (approved):
utACK 8dec9c560b53488c1e71d8f74241c7dce42cb387

Sjors commented:
cc @glozow & @instagibbs

sedited reviewed (approved):
Re-ACK 8dec9c560b53488c1e71d8f74241c7dce42cb387

glozow commented:
utACK 8dec9c560b

instagibbs reviewed (approved):
LGTM
